index.d.ts

```diff
@@ -18,23 +18,24 @@ import * as fastify from 'fastify'
 import * as http from 'http'
 
 import {FormatName} from 'ajv-formats'
+import {RouteGenericInterface} from 'fastify/types/route'
 
 export = customPlugin
 
-declare function customPlugin(envSchema?: customPlugin.environmentSchema): customPlugin.CustomService
+declare function customPlugin<ServiceConfig = NodeJS.Dict<string>>(envSchema?: customPlugin.environmentSchema): customPlugin.CustomService<ServiceConfig>
 
 declare namespace customPlugin {
-  type CustomService = (asyncInitFunction: AsyncInitFunction, serviceOptions?: CustomServiceOptions) => any
+  type CustomService<ServiceConfig = NodeJS.Dict<string>> = (asyncInitFunction: AsyncInitFunction<ServiceConfig>, serviceOptions?: CustomServiceOptions) => any
 
   function getDirectServiceProxy(serviceNameOrURL: string, options?: InitServiceOptions): Service
   function getServiceProxy(microserviceGatewayServiceName: string, options?: InitServiceOptions): Service
   interface environmentSchema {
     type: 'object',
-    required?: string[],
+    required?: string[] | readonly string[],
     properties: object
   }
 
-  type AsyncInitFunction = (service: DecoratedFastify) => Promise<void>
+  type AsyncInitFunction<ServiceConfig = NodeJS.Dict<string>> = (service: DecoratedFastify<ServiceConfig>) => Promise<void>
```

Comment:
The part ServiceConfig =... Is repeated multiple times.

Can you extract it in a common type?

Reply:
I've looked it up, but I can't find a way to extract a generic. It is not actually a type itself. 
Do you know how to do it?

Reply:
What about something like:

```
type ServiceConfig<T = NodeJS.Dict<string>> = T
```

?